**Preamble**

**Colloboration Policy**. The student is to *explicitly identify* his/her collaborators in the assignment. If the student did not work with anyone, he/she should indicate `Collaborators=['none']`. If the student obtains a solution through research (e.g., on the web), acknowledge the source, but *write up the solution in HIS/HER OWN WORDS*. There will be a one mark penalty if a student fails to indicate his/her collaborators.

**There will be NO EXCEPTIONS to this grading policy.**

Name = **ZHOU DUO**

Collaborators = **None,(There are two internet references, the URL is marked in the code block)**

In [2]:
# Marks 
# DO NOT CHANGE THIS

parts    = ['(a)    ', '(b.i)  ', '(b.ii) ', '(b.iii)', '(b.iv) ', '(b.v)  ', '(c)    ']
score    = [0,0,0,0,0,0,0]
maxscore = [4,2,1,3,2,3,5]

numparts = len(parts)

for i in range(numparts):
    print("score for {}: {} out of {}".format(parts[i], score[i], maxscore[i]))
print()
print("total            : {} out of {}".format(sum(score),sum(maxscore)))

score for (a)    : 0 out of 4
score for (b.i)  : 0 out of 2
score for (b.ii) : 0 out of 1
score for (b.iii): 0 out of 3
score for (b.iv) : 0 out of 2
score for (b.v)  : 0 out of 3
score for (c)    : 0 out of 5

total            : 0 out of 20


---

# Assignment 1 - Linear Programs with Many Optimum

In this assignment, you will:

(a) familiarize with the PuLP syntax and use PuLP to solve a linear programme (LP);

(b) determine when a LP has *alternative* optimal points and determine the set of *all optimal points*;

(c) familiarize with the PuLP syntax and use PuLP to solve larger linear programmes.


If you need help on using Jupyter notebooks, click <a href='#help'>here</a> for a simple guide.
A more thorough guide can be [here](https://www.codecademy.com/articles/how-to-use-jupyter-notebooks).

---

## Abby's Flying Fairy School

Abby's Flying Fairy School plans to sell some of Abby Caddaby's magical potions.
Abby has a catalogue of two potions *Amaze* and *Bedazzle*, that she plans to sell are \\$28 and \\$23 (per ml), respectively.

Each potion is made of 3 proprietary ingredients: Ingredient 1, Ingredient 2, and Ingredient 3, which cost \\$20, \\$22, and \\$23 (per ml), respectively. Of which, only 2021ml of Ingredient 1 and only 2021ml of Ingredient 2 are available.

To make the potions, Abby has to adhere to the following requirements.
- At least 40\% of *Amaze* has to be Ingredient 1;
- At most 20\% of *Bedazzle* has to be Ingredient 3.

Note that: $X_1$ ml of Ingredient 1, $X_2$ ml of Ingredient 2, and $X_3$ ml of Ingredient 3 will yield exactly $X_1+X_2+X_3$ ml of potion $X$. This is true for any (magic) potion.   


In summary, we has the following information.

| Potion | Amaze | Bedazzle |
| ---|---| ---|
| Revenue (per ml)     | \\$28 | \\$23|

| Ingredient | 1 | 2 | 3 |
| ---|---|---|---|
| Cost (per ml)  | \\$20 | \\$22 | \\$23 |
| Amount available | 2021ml | 2021 limit | no limit |

|Potion|Ingredient 1|Ingredient 2| Ingredient 3|
|---|---| ---| --- |
| Amaze    | $\ge 0.4$ | -- | -- |
| Bedazzle | -- | -- | $\le 0.2$ |

---

**(a) (4 marks)** You decide to use linear programming to help Abby maximize her profit. 
Formulate the linear programming problem and solve it using PuLP. 

To help you, I have listed six variables. You are to assign the variables some meaning.
- $A_1 = $ _Amout of Ingredient 1 in Amaze_ 
- $A_2 = $ _Amout of Ingredient 2 in Amaze_
- $A_3 = $ _Amout of Ingredient 3 in Amaze_
- $B_1 = $ _Amout of Ingredient 1 in Bedazzle_
- $B_2 = $ _Amout of Ingredient 2 in Bedazzle_
- $B_3 = $ _Amout of Ingredient 3 in Bedazzle_

So, it can be written as the fomula below:

$$  \begin{align*}
Max. \  & 8A_1+6A_2+5A_3+3B_1+B_2 \\
s.t. \  & -3A_1+2A_2+2A_3\leq 0 \\
 & -B_1-B_2+4B_3\ \ \ \ \leq 0 \\
 & A_1+B_1\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \  \ \ \ \ \leq 2021 \\
 & A_2+B_2\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \  \ \ \ \ \leq 2021 \\
 & A_1,A_2,A_3,B_1,B_2,B_3\geq 0 \\
\end{align*}   $$


In [5]:
# Method1 Form 1: Solve directly with pulp
from pulp import *

model = LpProblem("Abby_Potion", LpMaximize)

A1 = LpVariable('A1', lowBound=0)
A2 = LpVariable('A2', lowBound=0)
A3 = LpVariable('A3', lowBound=0)
B1 = LpVariable('B1', lowBound=0)
B2 = LpVariable('B2', lowBound=0)
B3 = LpVariable('B3', lowBound=0)

# WRITE YOUR CODE HERE

# Objective Function
model += 8.0*A1 + 6.0*A2 + 5.0*A3 + 3.0*B1 + B2 + 0*B3, "profit"

# s.t.
model += 3.0*A1-2.0*A2-2.0*A3 >= 0 , "A1 is Not less than 40%"

model += -B1-B2+4.0*B3 <= 0 , "B3 is Not more than 20%"

model += A1 + B1 <= 2021 , "constrain of Ingredient 1"

model += A2 + B2 <= 2021 , "constrain of Ingredient 2"

# write into lp file
model.writeLP("Abby_Potion.lp")

model.solve()

print("Status:", LpStatus[model.status])

for v in model.variables():
    print(v.name, "=", v.varValue)

print("Total Prefit of Abby = ", value(model.objective))


Status: Optimal
A1 = 2021.0
A2 = 0.0
A3 = 3031.5
B1 = 0.0
B2 = 2021.0
B3 = 0.0
Total Prefit of Abby =  33346.5


In [13]:
# Method1 Form2: Solve with pulp using Matirx
import pulp

# Objective coe
z = [8,6,5,3,1,0]

# A
a = [[-3,2,2,0,0,0], 
     [0,0,0,-1,-1,4],
     [1,0,0,1,0,0],
     [0,1,0,0,1,0]]

b = [0,0,2021,2021]

# define question
m = pulp.LpProblem(sense=pulp.LpMaximize)

# define x[1,2,3,4,5,6]
x = [pulp.LpVariable(f'x{i}', lowBound=0) for i in [1,2,3,4,5,6]]

# z %*% x Matrix
m += pulp.lpDot(z, x)

# Set conditions
for i in range(len(a)):
    m += (pulp.lpDot(a[i], x) <= b[i])
# solve
m.solve()
# print
print("Status:", LpStatus[m.status])

for v in m.variables():
    print(v.name, "=", v.varValue)

print("Total Prefit of Abby = ", value(m.objective))

Status: Optimal
x1 = 2021.0
x2 = 0.0
x3 = 3031.5
x4 = 0.0
x5 = 2021.0
x6 = 0.0
Total Prefit of Abby =  33346.5


In [14]:
# Method2 Sovle by Scipy.optimize
import numpy as np
from scipy import optimize as op

# bound
x1=(0,None)
x2=(0,None)
x3=(0,None)
x4=(0,None)
x5=(0,None)
x6=(0,None)

# Objective ceo
c = np.array([-8,-6,-5,-3,-1,0])

# A
A_ub = np.array([[-3,2,2,0,0,0], [0,0,0,-1,-1,4],[1,0,0,1,0,0],[0,1,0,0,1,0]])

# b
B_ub = np.array([0,0,2021,2021])

# solve
res=op.linprog(c,A_ub,B_ub,bounds=(x1,x2,x3,x4,x5,x6))

# fun:Optimal, x:value
print(res)

     con: array([], dtype=float64)
     fun: -33346.4999963016
 message: 'Optimization terminated successfully.'
     nit: 8
   slack: array([7.63020580e-09, 4.55532534e+02, 2.21557684e-07, 2.20891934e-07])
  status: 0
 success: True
       x: array([2.02100000e+03, 5.45135282e+02, 2.48636472e+03, 1.94283655e-09,
       1.47586472e+03, 2.55083046e+02])


3031.4

Optimal can be reached from the results of both forms. 

|A1 |A2 |A3 |B1 | B2 |B3 |Profit|
|---|---|---|---|---|---|---|
|2021|0|3031.5|0|2021|0|33346.5|
|2021|545|2486.4|0|1475.9|255|33346.5|

And it can be found that this problem has at least two solutions

Next, discuss the problem with adding slack variables.

---

**(b)** Even though PuLP did not state that the linear program has alternative optimum, we will show that the linear program is *degenerate* and has **alternative optimum**. We will also find the set of optimal solutions for this linear program.

**(b.i) (2 marks)** Introduce slack variables to the linear program in (a). Solve the new linear program using PuLP. As before, I have listed the four slack variables and hinted at the corresponding constraints.

In [125]:
# Method1 form1
import pulp as pulp
import numpy as np

slackmodel = pulp.LpProblem("Abby_Slack", pulp.LpMaximize)

A1 = pulp.LpVariable('A1', lowBound=0)
A2 = pulp.LpVariable('A2', lowBound=0)
A3 = pulp.LpVariable('A3', lowBound=0)
B1 = pulp.LpVariable('B1', lowBound=0)
B2 = pulp.LpVariable('B2', lowBound=0)
B3 = pulp.LpVariable('B3', lowBound=0)
s1 = pulp.LpVariable('s1', lowBound=0) # amount of Ingredient 1 available 
s2 = pulp.LpVariable('s2', lowBound=0) # amount of Ingredient 2 available 
s3 = pulp.LpVariable('s3', lowBound=0) # amount of Ingredient 1 in Potion A
s4 = pulp.LpVariable('s4', lowBound=0) # amount of Ingredient 3 in Potion B




# WRITE YOUR CODE HERE

# Objective Function
slackmodel += 8.0*A1 + 6.0*A2 + 5.0*A3 + 3.0*B1 + B2 + 0*B3, "profit"

# s.t.
slackmodel += A1 + B1 + s1 == 2021 , "I1_s"

slackmodel += A2 + B2 + s2 == 2021 , "I2_s"

slackmodel += -3.0 * A1 + 2.0 * A2 + 2.0 * A3 + s3 == 0 , "A1_s"

slackmodel += -B1 - B2 + 4.0 * B3 +s4 == 0 , "B3_s"


# write into lp file
slackmodel.writeLP("Abby_Slack.lp")

slackmodel.solve()

print("Status:", LpStatus[slackmodel.status])

for v in slackmodel.variables():
    print(v.name, "=", v.varValue)

print("Total Prefit of Abby = ", value(slackmodel.objective))


Status: Optimal
A1 = 2021.0
A2 = 0.0
A3 = 3031.5
B1 = 0.0
B2 = 2021.0
B3 = 505.25
s1 = 0.0
s2 = 0.0
s3 = 0.0
s4 = 0.0
Total Prefit of Abby =  33346.5


**(b.ii)(1 mark)** Identify the nonbasic and basic variables in the solution provided by PuLP.


**Write your answer to (b.ii) here.**

From the Method1 form1, we can get solution below:

**Basic Variables**: _A1, A3, B2, B3_

**Nonbasic Variables**: _A2, B1, s1, s2, s3, s4_

### **(b.iii)(3 marks)** Construct the simplex tableau corresponding to the basic variables you identified in **(b.ii)**.

If you need help on using `numpy` for matrix operations, click <a href='#matrix'>here</a>. 

In [126]:
# Compute the Matrix
list_var = [ str(v) for v in slackmodel.variables()]
list_constr = [ c for c in slackmodel.constraints]

print ("list of variables   : ", list_var)
print ("list of constraints : ", list_constr)
print()


constraint_dict = {(c, var): 0 for c in list_constr for var in list_var}
for (var, c, coeff) in slackmodel.coefficients():
    constraint_dict[(c,var)] = coeff
    
constraint_matrix = np.matrix([[constraint_dict[(c,var)] for var in list_var] for c in list_constr])
print("constraint matrix, A: ")
print(constraint_matrix)


obj_dict = { var: 0 for var in list_var}
for (var,coeff) in slackmodel.objective.items():
    obj_dict[str(var)] = coeff

first_row = np.matrix([obj_dict[var] for var in list_var])
print("objective vector, c: ")
print(first_row)

RHS = np.matrix([- slackmodel.constraints[c].constant for c in list_constr])
print("RHS , b:")
print(RHS)

cols = ['A1', 'A3', 'B2', 'B3']
cols2 = ['A2', 'B1', 's1', 's2', 's3', 's4']

print("B, {}:".format(cols))
B = np.matrix([[constraint_dict[(c,var)] for var in cols] for c in list_constr])
print(B)

print("N, {}:".format(cols2))
N = np.matrix([[constraint_dict[(c,var)] for var in cols2] for c in list_constr])
print(N)

print("cB, {}:".format(cols))
cB = np.matrix([obj_dict[var] for var in cols])
print(cB)

print("cN, {}:".format(cols))
cN = np.matrix([obj_dict[var] for var in cols2])
print(cN)

B_N = np.linalg.inv(B)*N
print('inv(B)*N',B_N)
B_b = np.linalg.inv(B)*RHS.T
print('inv(B)*b',B_b)
UP_l = cN - cB * np.linalg.inv(B) * N
print('UP_l',UP_l)
UP_r = -cB * np.linalg.inv(B)*RHS.T
print('UP_r',UP_r)


list of variables   :  ['A1', 'A2', 'A3', 'B1', 'B2', 'B3', 's1', 's2', 's3', 's4']
list of constraints :  ['I1_s', 'I2_s', 'A1_s', 'B3_s']

constraint matrix, A: 
[[ 1.  0.  0.  1.  0.  0.  1.  0.  0.  0.]
 [ 0.  1.  0.  0.  1.  0.  0.  1.  0.  0.]
 [-3.  2.  2.  0.  0.  0.  0.  0.  1.  0.]
 [ 0.  0.  0. -1. -1.  4.  0.  0.  0.  1.]]
objective vector, c: 
[[8. 6. 5. 3. 1. 0. 0. 0. 0. 0.]]
RHS , b:
[[2021. 2021.   -0.   -0.]]
B, ['A1', 'A3', 'B2', 'B3']:
[[ 1.  0.  0.  0.]
 [ 0.  0.  1.  0.]
 [-3.  2.  0.  0.]
 [ 0.  0. -1.  4.]]
N, ['A2', 'B1', 's1', 's2', 's3', 's4']:
[[ 0.  1.  1.  0.  0.  0.]
 [ 1.  0.  0.  1.  0.  0.]
 [ 2.  0.  0.  0.  1.  0.]
 [ 0. -1.  0.  0.  0.  1.]]
cB, ['A1', 'A3', 'B2', 'B3']:
[[8. 5. 1. 0.]]
cN, ['A1', 'A3', 'B2', 'B3']:
[[6. 3. 0. 0. 0. 0.]]
inv(B)*N [[ 0.    1.    1.    0.    0.    0.  ]
 [ 1.    1.5   1.5   0.    0.5   0.  ]
 [ 1.    0.    0.    1.    0.    0.  ]
 [ 0.25 -0.25  0.    0.25  0.    0.25]]
inv(B)*b [[2021.  ]
 [3031.5 ]
 [2021.  ]
 [ 505.2

In [1]:
# WORKING...
# It is a Reference code of Simplex Algerithm
# Refer to: https://zhuanlan.zhihu.com/p/60360871

import numpy as np


class Table:
    def __init__(self):
        pass

    def set_para(self, A, b, c, base, z0):
        self.A = A
        self.b = b
        self.c = c
        self.z0 = z0
        self.base = base
        self.m, self.n = self.A.shape
        # self.cj=np.nan
        self.x, self.y = self.c.shape

    def build(self):
        self.table = np.zeros((self.m + 1, self.n + 1))
        self.table[:-1, :1] = self.b.T
        self.table[-1, 0] = self.z0
        self.table[:-1, 1:] = self.A
        self.table[-1, 1:] = c
        self.baseVar = base

        self.table1 = np.zeros((1, self.n))
        # self.table1[0,0]=self.cj
        self.table1[:1, 0:] = self.c

        self.table2 = np.zeros((1, self.n), dtype=np.int)
        self.table2[:1, 0:] = list(range(1, int(self.n + 1)))

    def is_best(self):
        for sigma_index in range(self.n):
            if sigma_index not in self.baseVar:
                sigma = self.table[-1, 1 + sigma_index]
                if sigma > 0:
                    return False
        return True

    def is_no_solution(self):
        for sigma_index in range(self.n):
            if sigma_index not in self.baseVar:
                sigma = self.table[-1, 1 + sigma_index]
                if sigma > 0:
                    no_solution_flag = True
                    for a in self.table[:-1, 1 + sigma_index]:
                        if a > 0:
                            no_solution_flag = False
                    if no_solution_flag == True:
                        return True
        return False

    def get_inVar(self):
        max_sigma = 0
        inVar = None
        for sigma_index in range(self.n):
            if sigma_index not in self.baseVar:
                sigma = self.table[-1, 1 + sigma_index]
                if sigma > max_sigma:
                    max_sigma = sigma
                    inVar = sigma_index
        return inVar

    def get_outVar(self, inVar):
        rates = []
        rates1 = []
        for nobaseVar in range(self.m):
            a = self.table[nobaseVar, inVar + 1]
            b = self.table[nobaseVar, 0]
            if a > 0:
                rate = b / a
                rates.append([rate, nobaseVar])
            if a > 0:
                rate1 = round(b / a, 2)
                rates1.append([rate1, nobaseVar])
            else:
                rate1 = '/'
                rates1.append([rate1, nobaseVar])
        print('|theta|', end='')
        for theta in range(self.m):
            if isinstance(rates1[theta][0], str):
                print('%7s' % rate1, end=' |')
            else:
                print("%7.2f" % rates1[theta][0], end=' |')
        print()
        print()
        return min(rates)[1]

    def in_out(self, inVar, outVar):
        a = self.table[outVar, inVar + 1]
        self.table[outVar, :] /= a
        for i in range(self.m + 1):
            if i != outVar:
                self.table[i, :] -= self.table[outVar, :] * self.table[i, inVar + 1]
        self.baseVar[outVar] = inVar

    def show(self):
        print('Basic Variable/Value：', list(map(lambda n: n + 1, self.baseVar)), end='/')
        print(self.table[:-1, 0])
        print("Simplex Tableau")
        
        for x1 in range(self.x):
            print('|', end=' ')
            print('     ' + "Cj" + "      " + "|", end=' ')
            for y1 in range(self.y):
                print('%6.0f' % self.table1[x1, y1], end=' ')
                print('|', end=' ')
            print()        
        
        BASE = list(map(lambda n: n + 1, base))
        BASE.append('-z')
        for i in range(self.m + 1):
            print('|', end=' ')
            if type(BASE[i]) == int:
                print('%1s%-2d' % ('x', BASE[i]), end='')
            else:
                print(BASE[i], end='  ')
            if i != self.m:
                print('', end=' ')
            print('|', end=' ')
            for j in range(self.n + 1):
                print('%6.2f' % self.table[i, j], end=' ')
                print('|', end=' ')
            print()


        for x1 in range(self.x):

            print('|', end=' ')
            print('%-3s' % ("xb") + " " + "|", end='  ')
            print("Value", end=' ')
            print('|', end=' ')
            for y1 in range(self.y):
                print('    ' + 'x%d' % self.table2[x1, y1], end=' ')
                print('|', end=' ')
            # print('%6s'%'theta',end=' ')
            # aprint('|',end=' ')
            print()

        

class Simplex:
    def __init__(self):
        self.table = Table()

        # 0 正常，尚未得到最优解，继续迭代
        # 1 无解，无界解
        # 2 达到最优解
        self.status = 0
        self.inVar = None
        self.outVar = None

    def set_para(self, A, b, c, base, z0=0):
        self.table.set_para(A, b, c, base, z0)

    def output_result(self):
        self._main()
        if self.status == 1:
            print('=========')
            print("Unbound")
            print('=========')
        elif self.status == 2:
            print('=================')
            print("One Optimal")
            print('=================')
        elif self.status == 3:
            print('=====================')
            print("Infinite Optimal")
            print('=====================')

    def _main(self):
        self._build_table()
        while 1:
            self.table.show()
            if self._is_best() or self._is_no_solution():
                return

            self._get_inVar()
            self._get_outVar()
            self._in_out()

    def _build_table(self):
        self.table.build()

    def _is_best(self):
        if self.table.is_best():
            self.status = 2
            return True
        return False

    def _is_no_solution(self):
        if self.table.is_no_solution():
            self.status = 1
            return True
        return False

    def _get_inVar(self):
        self.inVar = self.table.get_inVar()

    def _get_outVar(self):
        self.outVar = self.table.get_outVar(self.inVar)

    def _in_out(self):
        self.table.in_out(self.inVar, self.outVar)

#################################################
if __name__ == "__main__":
    s = Simplex()
    A = np.matrix([[-3,2,2,0,0,0,1,0,0,0], 
                   [0,0,0,-1,-1,4,0,1,0,0],
                   [1,0,0,1,0,0,0,0,1,0],
                   [0,1,0,0,1,0,0,0,0,1]])

    b = np.matrix([[0,0,2021,2021]])
    c = np.matrix([[8,6,5,3,1,0,0,0,0,0]])
    base = [7,8,9,10]
    base = list(map(lambda n: n - 1, base))
    s.set_para(A, b, c, base)
    s.output_result()

Basic Variable/Value： [7, 8, 9, 10]/[   0.    0. 2021. 2021.]
Simplex Tableau
|      Cj      |      8 |      6 |      5 |      3 |      1 |      0 |      0 |      0 |      0 |      0 | 
| x7  |   0.00 |  -3.00 |   2.00 |   2.00 |   0.00 |   0.00 |   0.00 |   1.00 |   0.00 |   0.00 |   0.00 | 
| x8  |   0.00 |   0.00 |   0.00 |   0.00 |  -1.00 |  -1.00 |   4.00 |   0.00 |   1.00 |   0.00 |   0.00 | 
| x9  | 2021.00 |   1.00 |   0.00 |   0.00 |   1.00 |   0.00 |   0.00 |   0.00 |   0.00 |   1.00 |   0.00 | 
| x10 | 2021.00 |   0.00 |   1.00 |   0.00 |   0.00 |   1.00 |   0.00 |   0.00 |   0.00 |   0.00 |   1.00 | 
| -z  |   0.00 |   8.00 |   6.00 |   5.00 |   3.00 |   1.00 |   0.00 |   0.00 |   0.00 |   0.00 |   0.00 | 
| xb  |  Value |     x1 |     x2 |     x3 |     x4 |     x5 |     x6 |     x7 |     x8 |     x9 |     x10 | 
|theta|      / |      / |2021.00 |      / |

Basic Variable/Value： [7, 8, 1, 10]/[6063.    0. 2021. 2021.]
Simplex Tableau
|      Cj      |      8 |      6 |      

<ipython-input-1-939a68a403f1>:34: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  self.table2 = np.zeros((1, self.n), dtype=np.int)


**Fill in the table for (b.iii) here.**

$$ Original\ Tableau $$

| $Z$ | $A_1$ | $A_2$ | $A_3$ | $B_1$ | $B_2$ | $B_3$ | $s_1$ | $s_2$ | $s_3$ | $s_4$ | RHS |
|:----|:------|:------|:------|:------|:------|:------|:------|:------|:------|:------|:----|
| 1   | -8    | -6    | -5    | -3    | -1    | 0     | 0     | 0     | 0     | 0     | 0   |
| 0   | -3    | 2     | 2     | 0     | 0     | 0     | 1     | 0     | 0     | 0     | 0   |
| 0   | 0     | 0     | 0     | -1    | -1    | 4     | 0     | 1     | 0     | 0     | 0   |
| 0   | 1     | 0     | 0     | 1     | 0     | 0     | 0     | 0     | 1     | 0     | 2021|
| 0   | 0     | 1     | 0     | 0     | 1     | 0     | 0     | 0     | 0     | 1     | 2021|

### Three Tableau are same

$$ Basic\ Variable: A_1 = 2021, A_2 = 2021, A_3 = 1010.5, s_2 = 0 $$

| $Z$ | $A_1$ | $A_2$ | $A_3$ | $B_1$ | $B_2$ | $B_3$ | $s_1$ | $s_2$ | $s_3$ | $s_4$ | RHS |
|:----|:------|:------|:------|:------|:------|:------|:------|:------|:------|:------|:----|
| 1   | 0     | 0     | 0     | 12.5  | 0     | 0     | 2.5   | 0     | 15.5  | 1     | 33346.5|
| 0   | 0     | 0     | 1     | 1.5   | -1    | 0     | 0.5   | 0     | 1.5   | -1    |1010.5   |
| 0   | 0     | 0     | 0     | -1    | -1    | 4     | 0     | 1     | 0     | 0     | 0   |
| 0   | 1     | 0     | 0     | 1     | 0     | 0     | 0     | 0     | 1     | 0     | 2021|
| 0   | 0     | 1     | 0     | 0     | 1     | 0     | 0     | 0     | 0     | 1     | 2021|


$$ Basic\ Variable: A_1 = 2021, A_2 = 2021, A_3 = 1010.5, B_3 = 0 $$

| $Z$ | $A_1$ | $A_2$ | $A_3$ | $B_1$ | $B_2$ | $B_3$ | $s_1$ | $s_2$ | $s_3$ | $s_4$ | RHS |
|:----|:------|:------|:------|:------|:------|:------|:------|:------|:------|:------|:----|
| 1   | 0     | 0     | 0     | 12.5  | 0     | 0     | 2.5   | 0     | 15.5  | 1     | 33346.5|
| 0   | 0     | 0     | 1     | 1.5   | -1    | 0     | 0.5   | 0     | 1.5   | -1    |1010.5   |
| 0   | 0     | 0     | 0     | -0.25 | -0.25 | 1     | 0     | 0.25    | 0     | 0     | 0   |
| 0   | 1     | 0     | 0     | 1     | 0     | 0     | 0     | 0     | 1     | 0     | 2021|
| 0   | 0     | 1     | 0     | 0     | 1     | 0     | 0     | 0     | 0     | 1     | 2021|



$$Basic\ Variable: A_1 = 2021, A_3 = 3031.5, B_2 = 2021, B_3 = 505.25 $$

| $Z$ | $A_1$ | $A_2$ | $A_3$ | $B_1$ | $B_2$ | $B_3$ | $s_1$ | $s_2$ | $s_3$ | $s_4$ | RHS |
|:----|:------|:------|:------|:------|:------|:------|:------|:------|:------|:------|:----|
| 1   | 0     | 0     | 0     | 12.5 | 0     | 0     | 15.5 |1     | 2.5  | 0     | 33346.5|
| 0   | 1     | 0     | 0     | 1     | 0     | 0     | 1     | 0     | 0   | 0    | 2021   |
| 0   | 0     | 1     | 1     | 1.5   | 0     | 0     | 1.5   | 0     | 0.5     | 0     | 3031.5   |
| 0   | 0     | 1     | 0     | 0     | 1     | 0     | 0     | 1     | 1     | 0     | 2021|
| 0   | 0     | 0     | 0     | -0.25 | 0     | 1     | 0     | 0.25  | 0     | 0.25  | 505.25|

#### Choose a table arbitrarily(actually all are the same), we can get the relationship between the variables

$$A_1=2021, \\
A_2+A_3=3031.5,\to A_2=2021-B_2 \\
A_2+B_2=2021,\to A_3 = 1010.5+B_2\\
B_1=0\\
B_2=B_2\\
B_3=B_3\\
s_1,s_2,s_3=0\\
B_3+0.25s_4=505.25
$$

$$ \{(A_1,A_2,A_3,B_1,B_2,B_3,s_1,s_2,s_3,s_4)=(2021,2021-B_2,1010.5+B_2,0,B_2,B_3,0,0,0,2021-4B_3),0 \leq B_2\leq 2021,0\leq B_3\leq 505.25 \} $$

**(b.iv)(2 marks)** Using the table in (b.iii), explain why there is more than one optimal solution and determine the set of all optimal solutions for the linear program.

**Write your answer to (b.iv) here.**

* There are multiple solutions to this problem in linear programming. The algebraic interpretation is that the target vector is a linear combination of a few constraint vectors; the geometric interpretation is that the target vector is perpendicular to the optimal hyperplane.


* In the simplex table when the simplex method stops, if a non-basic variable has a coefficient of zero, then the non-basic variable will enter the basis without changing the objective function. These are multiple basic optimal solutions.


* You can get a new basic optimal solution by letting the variable into the base.


* For the new basic optimal solution, there may also be a situation where the coefficient of the non-basic variable is zero, and one of the solutions into the basic will get the solution just now;


* Other advances in the foundation will get a new basic optimal solution. Through this iterative search and eliminate all repeated basic optimal solutions, all basic optimal solution sets are obtained.

$$A_1=2021, \\
A_2+A_3=3031.5,\to A_2=2021-B_2 \\
A_2+B_2=2021,\to A_3 = 1010.5+B_2\\
B_1=0\\
B_2=B_2\\
B_3=B_3\\
s_1,s_2,s_3=0\\
B_3+0.25s_4=505.25
$$

$$ \{(A_1,A_2,A_3,B_1,B_2,B_3,s_1,s_2,s_3,s_4)=(2021,2021-B_2,1010.5+B_2,0,B_2,B_3,0,0,0,2021-4B_3),0 \leq B_2\leq 2021,0\leq B_3\leq 505.25 \} $$

**(b.v)(3 marks)** In addition to maximizing the profit, Abby also wants to *minimize cost*. Using your answer in (b.iv), determine the amounts of Amaze and Bedazzle (i.e. your answer should have *two* amounts) so that the *profit is the same as (a)* and the *cost is at most \\$108123.5*.

**Write your answer to (b.v) here.**

**Amount of Amaze**   : _5052.5_

**Amount of Bedazzle**: _0_

**Explanation**: _As below_



$$\begin{align*}
Min. \  & 20A_1+22A_2+23A_3+20B_1+22B_2+23B_3 \\
s.t. \  & 20A_1+22A_2+23A_3+20B_1+22B_2+23B_3\leq 108123.5 \\
 & A_1=2021, \\
 & A_2+B_2=2021 \\
 & A_3-B_2=1010.5, \\
 & B_2\leq2021 \\
 & B_3\leq505.25 \\
 & A_1,A_2,A_3,B_1,B_2,B_3\geq 0 \\
\end{align*}$$



$$Reduce\ to: \ \begin{align*}
Min. \  & 23B_2+23B_3 \\
s.t. \  & B_2\leq2021 \\
 & B_3\leq505.25 \\
 & B_2,B_3\geq 0 \\
\end{align*} $$

$$Solved: \ when\ B_2=B_3=0, \ Min.=0$$

$$ Namly: \ A_1=2021,\ A_2=2021,\ A_3=1010.5,\ cost_{min}=108123.5¥ $$

---
**(c) (5 marks) A New Venture.** One day, Abby discovers a warehouse of magical coffee beans along with a enchantment recipebook. 
There are **8 bean types** (Bean 0, Bean 1, ... Bean 7) and they can be mixed and roasted in certain proportions to produce **10 coffee blends** (Blend 0, Blend 1, ..., Blend 9).
The recipebook provides detailed instructions and the required amounts to produce one sack of coffee blend. In the cell below, we reproduce the mixing instructions, along with the available amount of beans.

In the cell below, we also provide the profit data, that is, the amount of profit that Abby can make from one sack of coffee blend (for each blend). However, there is a *special magical instruction*!
> For each Blend $i$ ($0\le i\le 9$), the number of sacks that you produce must be a perfect square!
> A number is a *perfect square* if it is the square of some integer. For example, the first 5 perfect squares are $0$, $1$, $4$, $9$, and $16$.

Your task is to determine the number of sacks to produce so that Abby can *maximize her profit*.
Remember that you have to adhere to the special magical instruction. 

Some notes:

- Full credit will only be awarded if a detailed explanation is given. Marks will *NOT* be awarded for **extensive** "trial and error" or "guess and check" or [its variants](https://en.wikipedia.org/wiki/Trial_and_error). Here, **extensive** means that you try for all possible "perfect squares", i.e. you try all solutions in $\{(X_0,X_1,\ldots,X_9): X_i\text{ is a perfect square for all } i\}$.

- **One mark** will be awarded for determining the maximum profit that Abby can make. <font color = 'red'>You to use `PuLP` to solve the linear program with **SLACK** variables.</font>

- **IMPORTANT**. Python, like other software, suffers from [floating point errors](https://www.geeksforgeeks.org/floating-point-error-in-python/). For example, if you evaluate `1.2-1.0`, you get the answer `0.19999999999999996`. Do not worry too much about this. In our course, if two values "differ" by less than 0.0001, you can assume that the two values are the same. So, in the example, you can just treat `0.19999999999999996` as `0.2`.

- If you need help on implementing *large* LP in `PuLP`, you can refer to this [Sudoku example](https://coin-or.github.io/pulp/CaseStudies/a_sudoku_problem.html).

- If you need help on extracting information from `PuLP` variables / classes, click <a href='#pulp_extract'>here</a>. 

- You may (most likely) use  more than one cell. Also, you may need a variety of *Code* or *Markdown* cells.

In [5]:
import numpy as np

# DO NOT CHANGE THE CODE
# Evaluate the cell to "parse" the data.

n = 10
m = 8
blends = [i for i in range(n)]
beans  = [i for i in range(m)] 

profit = [14, 16, 19, 8, 14, 11, 7, 7, 35, 30]
for i in range(n):
    print("Blend {} yields ${} per sack.".format(i,profit[i]))
print()

amt = [265, 338, 275, 488, 207, 276, 239, 117]
print("Beans Availability:")
for i in range(m):
    print("Amount of Ingredient {} Available = {} bottles.".format(i,amt[i]))



mixing = np.matrix([
    [2, 2, 0, 2, 1, 1, 0, 0, 5, 2] ,
    [2, 3, 2, 1, 1, 1, 1, 0, 0, 0] ,
    [1, 1, 3, 0, 1, 1, 1, 1, 2, 2] ,
    [2, 3, 2, 2, 2, 2, 1, 1, 1, 2] ,
    [0, 1, 2, 0, 1, 1, 1, 0, 3, 5] ,
    [1, 1, 1, 1, 2, 1, 0, 1, 4, 4] ,
    [1, 1, 2, 0, 1, 1, 1, 1, 2, 5] ,
    [1, 0, 1, 0, 1, 0, 0, 1, 4, 2] ,
])


print("\nRecipe Instructions:")
for j in range(10):
    print("\nBlend{} requires:".format(j))
    for i in range(8):
        print("{} bottle(s) of Bean{}".format(mixing[i,j],i), end = ", ")
        if i in [3,7]: print() 


Blend 0 yields $14 per sack.
Blend 1 yields $16 per sack.
Blend 2 yields $19 per sack.
Blend 3 yields $8 per sack.
Blend 4 yields $14 per sack.
Blend 5 yields $11 per sack.
Blend 6 yields $7 per sack.
Blend 7 yields $7 per sack.
Blend 8 yields $35 per sack.
Blend 9 yields $30 per sack.

Beans Availability:
Amount of Ingredient 0 Available = 265 bottles.
Amount of Ingredient 1 Available = 338 bottles.
Amount of Ingredient 2 Available = 275 bottles.
Amount of Ingredient 3 Available = 488 bottles.
Amount of Ingredient 4 Available = 207 bottles.
Amount of Ingredient 5 Available = 276 bottles.
Amount of Ingredient 6 Available = 239 bottles.
Amount of Ingredient 7 Available = 117 bottles.

Recipe Instructions:

Blend0 requires:
2 bottle(s) of Bean0, 2 bottle(s) of Bean1, 1 bottle(s) of Bean2, 2 bottle(s) of Bean3, 
0 bottle(s) of Bean4, 1 bottle(s) of Bean5, 1 bottle(s) of Bean6, 1 bottle(s) of Bean7, 

Blend1 requires:
2 bottle(s) of Bean0, 3 bottle(s) of Bean1, 1 bottle(s) of Bean2, 3 bott

## Remarks on Answer

The solution given below essentially follows the steps in (a) and (b). 
However, most parts are automated as the number of variables and constraints have increased.

### The mathematical model can be obtained as follows

$$\begin{align*}
Max. \  & a \cdot Blends_i , \ \ i\in \{ 0,1,2,3,4,5,6,7,8,9 \} \\
s.t. \  & A_{8*18}\cdot Beens_i\leq b , \ \  i\in \{0,1,2,3,4,5,6,7\}  \\
 & Blends_i \in  \{ 1^2,2^2,3^2,\ddots ,24^2\} \\
\end{align*}$$

### Increase slack variable 

In [1]:
import pulp
#from functools import reduce

m = pulp.LpProblem("Abby_Slack_Coffee",LpMaximize)

# WRITE YOUR CODE HERE
# Objective Function
z = [14, 16, 19, 8, 14, 11, 7, 7, 35, 30, 0, 0, 0, 0, 0, 0, 0, 0]

# Strains
a = [[2, 2, 0, 2, 1, 1, 0, 0, 5, 2, 1, 0, 0, 0, 0, 0, 0, 0] ,
    [2, 3, 2, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0] ,
    [1, 1, 3, 0, 1, 1, 1, 1, 2, 2, 0, 0, 1, 0, 0, 0, 0, 0] ,
    [2, 3, 2, 2, 2, 2, 1, 1, 1, 2, 0, 0, 0, 1, 0, 0, 0, 0] ,
    [0, 1, 2, 0, 1, 1, 1, 0, 3, 5, 0, 0, 0, 0, 1, 0, 0, 0] ,
    [1, 1, 1, 1, 2, 1, 0, 1, 4, 4, 0, 0, 0, 0, 0, 1, 0, 0] ,
    [1, 1, 2, 0, 1, 1, 1, 1, 2, 5, 0, 0, 0, 0, 0, 0, 1, 0] ,
    [1, 0, 1, 0, 1, 0, 0, 1, 4, 2, 0, 0, 0, 0, 0, 0, 0, 1] ]

b = [265, 338, 275, 488, 207, 276, 239, 117]

c = [i**2 for i in range(24)]

# define x[10] as Beens
x = [pulp.LpVariable(f'x{i}', 0, None, LpInteger) for i in range(18)]

# z %*% x Matrix
m += pulp.lpDot(z, x)

# Set conditions
for i in range(len(a)):
#    if i == 1:
#        m += (pulp.lpDot(a[i], x) >= b[i])
#    else:
    m += (pulp.lpDot(a[i], x) <= b[i])
#    m += reduce(lambda x,y:x * y,c-x[i]) == 0
# solve
m.solve()

#print(m) 

# print optimal & variables
print("Status:", LpStatus[m.status])

for v in m.variables():
    print(v.name, "=", v.varValue)

print("Total Prefit of Abby = ", value(m.objective))



NameError: name 'LpMaximize' is not defined

#### Basic Variable:  $ X_0,X_1,X_2,X_3,X_4,X_5,X_6,X_7,X_8 $

In [124]:
list_var = [ str(v) for v in m.variables()]
list_constr = [ c for c in m.constraints]

print ("list of variables   : ", list_var)
print ("list of constraints : ", list_constr)
print()


constraint_dict = {(c, var): 0 for c in list_constr for var in list_var}
for (var, c, coeff) in m.coefficients():
    constraint_dict[(c,var)] = coeff
    
constraint_matrix = np.matrix([[constraint_dict[(c,var)] for var in list_var] for c in list_constr])
print("constraint matrix, A: ")
print(constraint_matrix)


obj_dict = { var: 0 for var in list_var}
for (var,coeff) in m.objective.items():
    obj_dict[str(var)] = coeff

first_row = np.matrix([obj_dict[var] for var in list_var])
print("objective vector, c: ")
print(first_row)

RHS = np.matrix([- m.constraints[c].constant for c in list_constr])
print("RHS , b:")
print(RHS)

cols = ['x0', 'x1', 'x2', 'x3', 'x4', 'x5', 'x6', 'x7']
cols2 = ['x8','x9', 'x10', 'x11', 'x12', 'x13', 'x14', 'x15', 'x16', 'x17']

print("B, {}:".format(cols))
B = np.matrix([[constraint_dict[(c,var)] for var in cols] for c in list_constr])
print(B)

print("N, {}:".format(cols2))
N = np.matrix([[constraint_dict[(c,var)] for var in cols2] for c in list_constr])
print(N)

print("cB, {}:".format(cols))
cB = np.matrix([obj_dict[var] for var in cols])
print(cB)

print("cN, {}:".format(cols))
cN = np.matrix([obj_dict[var] for var in cols2])
print(cN)

B_N = np.linalg.inv(B)*N
print('DW_l',B_N)
B_b = np.linalg.inv(B)*RHS.T
print('DW_r',B_b)
UP_l = cN - cB * np.linalg.inv(B) * N
print('UP_l',UP_l)
UP_r = -cB * np.linalg.inv(B)*RHS.T
print('UP_r',UP_r)

list of variables   :  ['x0', 'x1', 'x10', 'x11', 'x12', 'x13', 'x14', 'x15', 'x16', 'x17', 'x2', 'x3', 'x4', 'x5', 'x6', 'x7', 'x8', 'x9']
list of constraints :  ['_C1', '_C2', '_C3', '_C4', '_C5', '_C6', '_C7', '_C8']

constraint matrix, A: 
[[2 2 1 0 0 0 0 0 0 0 0 2 1 1 0 0 5 2]
 [2 3 0 1 0 0 0 0 0 0 2 1 1 1 1 0 0 0]
 [1 1 0 0 1 0 0 0 0 0 3 0 1 1 1 1 2 2]
 [2 3 0 0 0 1 0 0 0 0 2 2 2 2 1 1 1 2]
 [0 1 0 0 0 0 1 0 0 0 2 0 1 1 1 0 3 5]
 [1 1 0 0 0 0 0 1 0 0 1 1 2 1 0 1 4 4]
 [1 1 0 0 0 0 0 0 1 0 2 0 1 1 1 1 2 5]
 [1 0 0 0 0 0 0 0 0 1 1 0 1 0 0 1 4 2]]
objective vector, c: 
[[14 16  0  0  0  0  0  0  0  0 19  8 14 11  7  7 35 30]]
RHS , b:
[[265 338 275 488 207 276 239 117]]
B, ['x0', 'x1', 'x2', 'x3', 'x4', 'x5', 'x6', 'x7']:
[[2 2 0 2 1 1 0 0]
 [2 3 2 1 1 1 1 0]
 [1 1 3 0 1 1 1 1]
 [2 3 2 2 2 2 1 1]
 [0 1 2 0 1 1 1 0]
 [1 1 1 1 2 1 0 1]
 [1 1 2 0 1 1 1 1]
 [1 0 1 0 1 0 0 1]]
N, ['x8', 'x9', 'x10', 'x11', 'x12', 'x13', 'x14', 'x15', 'x16', 'x17']:
[[5 2 1 0 0 0 0 0 0 0]
 [0 0 0 1 0 0 0 

### Explain

#### We fill out the results and get the tableau. The first row of $x_8$ is $0$, and the first row of $x_9$~$x_{17}$ is all non-zero real numbers.

| $Z$ | $x_1\ to\ x_7$ | $x_8$ | $x_9\ to\ x_{17} $ | $RHS$ |
|:----:|:------|:------|:------|:------|
| 1   | 0              | 0     | ~                  | 3069 |
| 0   | $I$            | 6     | ~                  | 16     | 
|     |                | -8    | ~                  | 25   |  
|     |                | 1     | ~                  | 36    | 
|     |                | 1     |  ~                 | 49 | 
|     |                | 5     |   ~                | 49  |
|     |                | 2     |    ~               | 36     | 
|     |                | 4     |     ~              | 25   |  
|     |                | -7    |      ~             | 16     | 

####  Let $x_9\ to\ x_{17} $ take the value 0 so that Z takes the maximum value. Found that using $x_8$ can express all other variables

$$ Z=3069,\\
x_0=16-6x_8,\\
x_1=25+8x_8,\\
x_2=36-x_8,\\
x_3=49-x_8,\\
x_4=49-5x_8,\\
x_5=36-2x_8,\\
x_6=25-4x_8,\\
x_7=16+7x_8,\\
x_8=x_8,\\
x_i=0,i\in \{9,10,11,12,13,14,15,16,17\} $$

#### So


$$ \{(x_0,x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_{10},x_{11},x_{12},x_{13},x_{14},x_{15},x_{16},x_{17})=(16-6x_8,25+8X_8,36-x_8,49-x_8,49-5x_8,36-2x_8,25-4x_8,16+7x_8,x_8,0,0,0,0,0,0,0,0,0),x_8 \in \{0,1,2\} \} $$

#### When X8 is equal to 0, the optimal solution can satisfy the condition

$$ \{(x_0,x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_{10},x_{11},x_{12},x_{13},x_{14},x_{15},x_{16},x_{17})=(16,25,36,49,49,36,25,16,0,0,0,0,0,0,0,0,0,0), \} $$

In [130]:
#Sudoku method verification
#refer to https://coin-or.github.io/pulp/CaseStudies/a_sudoku_problem.html

from pulp import *

model_coffee = LpProblem("Abby_Coffee", LpMaximize)

# WRITE YOUR CODE HERE

nrange = range(10)
mrange = range(8)
krange = range(24)

blends = [i for i in nrange]
beans  = [i for i in mrange] 
packs = [i for i in krange]

profit = [14, 16, 19, 8, 14, 11, 7, 7, 35, 30]

amt = [265, 338, 275, 488, 207, 276, 239, 117]

mixing = [[2, 2, 0, 2, 1, 1, 0, 0, 5, 2] ,
    [2, 3, 2, 1, 1, 1, 1, 0, 0, 0] ,
    [1, 1, 3, 0, 1, 1, 1, 1, 2, 2] ,
    [2, 3, 2, 2, 2, 2, 1, 1, 1, 2] ,
    [0, 1, 2, 0, 1, 1, 1, 0, 3, 5] ,
    [1, 1, 1, 1, 2, 1, 0, 1, 4, 4] ,
    [1, 1, 2, 0, 1, 1, 1, 1, 2, 5] ,
    [1, 0, 1, 0, 1, 0, 0, 1, 4, 2] ,]

# Sodoku 0 & 1
model_coffee = LpProblem("Abby_Coffee", LpMaximize)
choices = LpVariable.dicts("Choice", (nrange, krange), cat="Binary")

# Objective Function
model_coffee += lpSum( [j * j* profit[i] * choices[i][j] for j in krange for i in nrange]) 

# Constrains
for i in nrange:
    model_coffee += lpSum([choices[i][j] for j in krange]) <= 1
for i in mrange:
    model_coffee += lpSum([ j * j* mixing[i][m]* choices[m][j]  for j in krange for m in nrange]) <= amt[i]

# Solve
model_coffee.solve()

# Check solution
print("Status:", LpStatus[model_coffee.status])
for v in model_coffee.variables():
    if v.varValue == 1:
        print(f'The number of x{v.name[-3]} coffee blend is {int(v.name[-1])**2}')
print("Profit", format(model_coffee.objective.value()))

Status: Optimal
The number of x0 coffee blend is 16
The number of x1 coffee blend is 25
The number of x2 coffee blend is 36
The number of x3 coffee blend is 49
The number of x4 coffee blend is 49
The number of x5 coffee blend is 36
The number of x6 coffee blend is 25
The number of x7 coffee blend is 16
Profit 3069.0


## Summary

### The optimal goal can be achieved by both the matrix method and the Sudoku method



$$ \{(x_0,x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_{10},x_{11},x_{12},x_{13},x_{14},x_{15},x_{16},x_{17})=(16-6x_8,25+8X_8,36-x_8,49-x_8,49-5x_8,36-2x_8,25-4x_8,16+7x_8,x_8,0,0,0,0,0,0,0,0,0),x_8 \in \{0,1,2\} \} $$


---
<a id='help'></a>

### Using Jupyter Notebooks

When you click to the left of this box, you will notice that this box is highlighted by a slighly larger box. This is a *cell*. 

There are three types of cells in a notebook.

1. Markdown.
2. Code.
3. Raw.

You can change the type of cell by going to the tool bar.

You can *evaluate* cells by hitting **Shift+Enter**. Depending on the type of cells, you will have different outputs.

---

This is a **markdown** cell. Markdown is a lightweight markup language is similar to *html* with significantly less functionalities. However, the syntax is much simpler. You can find a [Markdown Cheatsheet here](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet).

---

In [9]:
# This is a CODE cell.
# After you hit Shift+Enter, it evaluates the cell in Python.
# Take note that in Python, to comment lines, you use the symbol #

print("Hello World!")

Hello World!


**Answering Questions**. You may choose to use *raw* or *markdown* cells to answer the questions. Of course, if the answer requires you to run a routine in Python, please use a *code* cell.

----

### Using `numpy` for matrix operations

<a id='matrix'></a>


Suppose we want to compute $$U^{-1}V$$
where
$$ 
U =\left(
\begin{array}{ccc}
1 & -6 & 0\\
0 & 2 & 1\\
0 & 2 & 0\\
\end{array}
\right)
\quad
V =\left(
\begin{array}{ccc}
1  &  3 &  0\\
16 & -3 &  0\\
11 & -1 & -1\\
\end{array}
\right)\,.
$$

The syntax is as follows.

In [92]:
import numpy as np # REMEMBER TO INCLUDE THIS LINE!

U = np.matrix([
    [1,-6,0],
    [0, 2,1],
    [0, 2,0],
]
)

V = np.matrix([
    [ 0, 3, 0],
    [16,-3, 0],
    [11,-1,-1],
]
)

M = np.linalg.inv(U)*V
print("inv(U)*V =")
print(np.linalg.inv(U))
print(M)

inv(U)*V =
[[ 1.   0.   3. ]
 [ 0.   0.   0.5]
 [-0.   1.  -1. ]]
[[33.   0.  -3. ]
 [ 5.5 -0.5 -0.5]
 [ 5.  -2.   1. ]]


---

### Extracting Information from `PuLP`

<a id='pulp_extract'></a>

Suppose we implemented the following linear program in `PuLP`.

```python
import numpy as np
import pulp as pulp

small_model = pulp.LpProblem("Brewer_Problem", pulp.LpMaximize)

Al = pulp.LpVariable('Ale', lowBound = 0)
Be = pulp.LpVariable('Beer', lowBound = 0)
S1 = pulp.LpVariable('S1', lowBound = 0)
S2 = pulp.LpVariable('S2', lowBound = 0)
S3 = pulp.LpVariable('S3', lowBound = 0)

small_model += 13*Al + 23*Be, "profit"

small_model += 5*Al  + 15*Be + S1 == 480, "corn" 
small_model += 4*Al  + 4*Be  + S2 == 160, "hops" 
small_model += 35*Al + 20*Be + S3 == 1190, "malt"
```

You can use the following syntax to find the constraint matrix $A$, the objective vector $c$, the RHS vector $b$, instead of writing by hand.

In [82]:
import numpy as np
import pulp as pulp

small_model = pulp.LpProblem("Brewer_Problem", pulp.LpMaximize)

Al = pulp.LpVariable('Ale', lowBound = 0)
Be = pulp.LpVariable('Beer', lowBound = 0)
S1 = pulp.LpVariable('S1', lowBound = 0)
S2 = pulp.LpVariable('S2', lowBound = 0)
S3 = pulp.LpVariable('S3', lowBound = 0)

small_model += 13*Al + 23*Be, "profit"

small_model += 5*Al  + 15*Be + S1 == 480, "corn" 
small_model += 4*Al  + 4*Be  + S2 == 160, "hops" 
small_model += 35*Al + 20*Be + S3 == 1190, "malt"

print(small_model)

print("****************")

list_var = [ str(v) for v in small_model.variables()]
list_constr = [ c for c in small_model.constraints]

print ("list of variables   : ", list_var)
print ("list of constraints : ", list_constr)
print()


constraint_dict = {(c, var): 0 for c in list_constr for var in list_var}
for (var, c, coeff) in small_model.coefficients():
    constraint_dict[(c,var)] = coeff
    
constraint_matrix = np.matrix([[constraint_dict[(c,var)] for var in list_var] for c in list_constr])
print("constraint matrix, A: ")
print(constraint_matrix)


obj_dict = { var: 0 for var in list_var}
for (var,coeff) in small_model.objective.items():
    obj_dict[str(var)] = coeff

first_row = np.matrix([obj_dict[var] for var in list_var])
print("objective vector, c: ")
print(first_row)

RHS = np.matrix([- small_model.constraints[c].constant for c in list_constr])
print("RHS , b:")
print(RHS)


print()
cols = ['Ale', 'S1', 'S3']
print("constraint matrix restricted to columns, {}:".format(cols))
B = np.matrix([[constraint_dict[(c,var)] for var in cols] for c in list_constr])
print(B)
print("first row restricted to columns, {}:".format(cols))
cB = np.matrix([obj_dict[var] for var in cols])
print(cB)
print("Status:", LpStatus[small_model.status])

for v in m.variables():
    print(v.name, "=", v.varValue)

print("Total Prefit of Abby = ", value(small_model.objective))

Brewer_Problem:
MAXIMIZE
13*Ale + 23*Beer + 0
SUBJECT TO
corn: 5 Ale + 15 Beer + S1 = 480

hops: 4 Ale + 4 Beer + S2 = 160

malt: 35 Ale + 20 Beer + S3 = 1190

VARIABLES
Ale Continuous
Beer Continuous
S1 Continuous
S2 Continuous
S3 Continuous

****************
list of variables   :  ['Ale', 'Beer', 'S1', 'S2', 'S3']
list of constraints :  ['corn', 'hops', 'malt']

constraint matrix, A: 
[[ 5 15  1  0  0]
 [ 4  4  0  1  0]
 [35 20  0  0  1]]
objective vector, c: 
[[13 23  0  0  0]]
RHS , b:
[[ 480  160 1190]]

constraint matrix restricted to columns, ['Ale', 'S1', 'S3']:
[[ 5  1  0]
 [ 4  0  0]
 [35  0  1]]
first row restricted to columns, ['Ale', 'S1', 'S3']:
[[13  0  0]]
Status: Not Solved
x0 = 4.0
x1 = 41.0
x10 = 0.0
x11 = 0.0
x12 = 0.0
x13 = 0.0
x14 = 0.0
x15 = 0.0
x16 = 0.0
x17 = 0.0
x2 = 36.0
x3 = 47.0
x4 = 39.0
x5 = 32.0
x6 = 17.0
x7 = 30.0
x8 = 2.0
x9 = 0.0
Total Prefit of Abby =  None
